martinize2/processors/repair_graph.py

```diff
@@ -131,12 +133,87 @@ def make_reference(mol):
+    # Rebuild missing atoms and canonicalize atomnames
+    missing = []
+    # Step 1: find all missing atoms. Canonicalize names while we're at it.
+    reference = ref_residue['reference']
+    found = ref_residue['found']
+    # found.graph['canonized'] = True
+    match = ref_residue['match']
+
+    resid = ref_residue['resid']
+    resname = ref_residue['resname']
+
+    for ref_idx in reference:
+        if ref_idx in match:
+            res_idx = match[ref_idx]
+            node = molecule.nodes[res_idx]
+            # Copy, because it's references everywhere.
+            node['graph'] = molecule.subgraph([res_idx]).copy()
+            node.update(reference.nodes[ref_idx])
+        else:
+#            if reference.nodes[ref_idx]['element'] != 'H':
+            print('Missing atom {}{}:{}'.format(resname, resid, reference.nodes[ref_idx]['atomname']))
+            missing.append(ref_idx)
+    # Step 2: try to add all missing atoms one by one. As long as we added
+    # *something* the situation changed, and we might be able to place another.
+    # We can only place atoms for which we know a neighbour.
+    added = True
+    while missing and added:
+        added = False
+        for ref_idx in missing:
+            # See if the atom we want to add has a neighbour for which we know
+            # the position. Otherwise, continue to the next.
+            if all(ref_neighbour in missing for ref_neighbour in reference[ref_idx]):
+                continue
+            added = True
+            missing.pop(missing.index(ref_idx))
+            res_idx = max(molecule) + 1  # Alternative: find the first unused number
+
+            # Create the new node
+            match[ref_idx] = res_idx
+            molecule.add_node(res_idx)
+            node = molecule.nodes[res_idx]
+            for key, val in ref_residue.items():
+                # Some attributes are only relevant on a residue level, not on
+                # an atom level.
+                if key not in ('match', 'found', 'reference'):
+                    node[key] = val
+            node.update(reference.nodes[ref_idx])
+
+            found.add_node(res_idx, **node)
+            # print("Adding {}{}:{}".format(resname, resid, node['atomname']))
+
+            neighbours = 0
+            for neighbour_ref_idx in reference[ref_idx]:
+                try:
+                    neighbour_res_idx = match[neighbour_ref_idx]
+                except KeyError:
+                    continue
+                if not molecule.has_edge(neighbour_res_idx, res_idx):
+                    molecule.add_edge(neighbour_res_idx, res_idx)
+                    neighbours += 1
+            assert neighbours != 0
+    if missing:
+        for ref_idx in missing:
+            print('Could not reconstruct atom {}{}:{}'.format(reference.nodes[ref_idx]['resname'],
+                  reference.nodes[ref_idx]['resid'], reference.nodes[ref_idx]['atomname']))
+
+
+def repair_graph(molecule, reference_graph):
```

Comment:
The docstring mentions a copy, but I do not see where it happens. Also, since the "molecule" parameter is a `Molecule`, then the return should be a `Molecule` too rather than just a `Graph`.

Reply:
And the docstring does not mention the PTMs.

Reply:
I just found the copy. It seems to be in the processor and not in this function.